### DISCUSSION TITLE
Workaround for some cases of the CORS problem - a simple Ajax proxy

### DISCUSSION THREAD

**Discussion by pgorod** [arendst/Tasmota#13818] [Show and tell]
So Tasmota had a few problems getting the CORS implementation right, and it's still not entirely fixed, so I developed a workaround which works great for my case, I leave it here in case it helps someone else.

I basically have a web server running in my local network, the same network that has all my Tasmota devices. This web server is accessible from the Internet, and using it to make Ajax requests to the devices would cause the browser to complain about the CORS issues.

So what I did was **ensure all my web requests go just to the web server, and the web server redirects some of those, where appropriate, to the devices**.

I created a PHP file on the server called **`ajaxproxy.php`**

```php
<?php

$deviceIPs = [
   'Cylinder'   => '192.168.1.85',
   'Heater'   => '192.168.1.88',
   'Circulator' => '192.168.1.84',
   'Valve'    => '192.168.1.90',
   'Garage'    => '192.168.1.87',
];

$url = $_GET['url'];
$url = 'http://' . $deviceIPs[$_GET['device']] .'/cm';
$ch = curl_init($url);

$params = array('cmnd' => $_GET['cmnd']);
$url = $url . '?' . http_build_query($params);
curl_setopt($ch, CURLOPT_URL, $url);
curl_setopt($ch, CURLOPT_HTTPGET, true);
curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
$response_json = curl_exec($ch);
curl_close($ch);

header('Content-type: application/json');
echo $response_json;
?>
```
Initially I had this simply redirecting the entire URL with the device and the commands in it, but I evolved to that version I posted because it makes the calling code look cleaner.

Then my plain Javascript Ajax client calls it like this:

```js
<script type="text/javascript">
function Ajax2Device(device, cmnd, path, ok_callback) {
    var xmlhttp = new XMLHttpRequest();
    xmlhttp.onreadystatechange = function() {
        if (xmlhttp.readyState == XMLHttpRequest.DONE) {   // XMLHttpRequest.DONE == 4
           if (xmlhttp.status == 200) {
               var data = JSON.parse(xmlhttp.responseText);
               var val = path.reduce((o, n) => o[n], data);
               ok_callback(val);
           }
           else if (xmlhttp.status == 400) {
              alert('There was an error 400');
           }
           else {
               alert('something else other than 200 was returned');
           }
        }
    };
    var encodedURL="ajaxproxy.php?cmnd=" + cmnd + "&device=" + device;
    xmlhttp.open("GET", encodedURL, true);
    xmlhttp.send();
}

function callit() {
    var data=Ajax2Device('Valve', 'status', ["Status", "Power"],
        function (val) {
          document.getElementById("ValveLight").innerHTML = val ? 'On' : 'Off';
    });
}
</script>
```

**`Ajax2Device`** is a generic function for all calls, it gets the following arguments:
- device name (which needs to be in `$deviceIPs` on the server side), 
- the Tasmota command (like `power 1` or whatever), 
- an array with the parts of the JSON response you're interested in, 
- and a callback function to execute asynchronously when the call returns, if successful.

**`callit()`** is just an example call to Ajax2Device so you see how it can be used.

No error handling, and quite careless programming, but you get the point, and well, it's a start.

This solves all CORS problems (with minor performance penalties) for this particular architecture. It won't solve everybody's problems, of course. I hope this can be useful to someone.

**Comment by thibaultferrier** [comment]
Thank you I will try something like this.
